tests/test_parser.py

```diff
@@ -131,3 +134,121 @@ def test_string_utf8():
     parser.consume_byte(b[:1])
     assert parser.valid_next_bytes() == set([Byte(b[1:])])
     parser.consume_byte(b[1:])
+
+
+class TestRecursiveNullableGrammars:
+    """
+    Computing parse tree of recursive nullable grammars will cause an infinite
+    loop if not handled correctly
+    """
+
+    @pytest.mark.timeout(5)
+    def test_no_infinite_loop(self):
+        """
+        A -> A
+        A ->
+
+        Loop occurs because `A -> A` is a nullable rule
+        """
+        # Note that we get a different grammar if we made `A = select([''], recurse=True)`
+        A = Select([], recursive=True)
+        A.values = [A, ""]
+        parser = EarleyCommitParser(A)
+        # Test that computing the parse tree doesn't hang
+        parser.parse_tree()
+        # Test that getting captures doesn't hang
+        parser.get_captures()
```

Comment:
Can you `assert` anything about the last two operations?

Also, if either of them can hang on an infinite loop, would it be worth having two tests, to identify which one hit the loop (although that may be more than needed).

Reply:
I reserved making asserts for some of the other tests, but I'll have a think :smile:

`get_captures` actually calls `parse_tree` under the hood BUT it can also hang even if `parse_tree` doesn't. I can maybe put another mock in here to disentangle them a bit..?

Reply:
I wouldn't worry about injecting another mock. I was just wondering if there was any visible state change which could be asserted after each step?